File: C/Cap03/c03ex06.c
// c03ex06.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char* inverte(const char* texto) {
  int tamanho = strlen(texto);
  char* invertido = malloc((tamanho + 1) * sizeof(char));
  for (int i = 0; i <= tamanho - 1; i++) {
    invertido[i] = texto[tamanho - i - 1];
  }
  invertido[tamanho] = '\0';
  return invertido;
}

char* trocar(const char* texto, int posicao, char novoCaractere) {
  int tamanho = strlen(texto);
  char* novoTexto = malloc((tamanho + 1) * sizeof(char));
  strcpy(novoTexto, texto);
  novoTexto[posicao - 1] = novoCaractere;
  return novoTexto;
}

int main(void) {
  char palavra[] = "COMPUTADOR";
  int tamanho = strlen(palavra);
  char copias[100];
  char* novaPalavra = trocar(palavra, 7, 'O');

  printf("Palavra ........: %s\n", palavra);
  printf("\n");
  printf("Lado esquerdo ..: %.3s\n", palavra);
  printf("Centro .........: %.4s\n", palavra + tamanho / 2 - 2);
  printf("Lado direito ...: %.3s\n", palavra + tamanho - 3);
  printf("Reverso ........: %s\n", inverte(palavra));
  strcpy(copias, palavra);
  strcat(copias, palavra);
  strcat(copias, palavra);
  printf("Copias .........: %s\n", copias);
  printf("Nova palavra ...: %s\n", novaPalavra);
  
  printf("\n");
  printf("Tecle <Enter> para encerrar... ");
  while (getchar() != '\n');

  free(novaPalavra);
  return 0;
}

File: GO/Cap03/c03ex06.go
// c03ex06.go

package main

import (
  "fmt"
  "strings"
)

func Inverte(texto string) string {
  var (
    caractere = []rune(texto)
    x rune
    i int
    j int
  )
  for i = 0; i <= (len(caractere) / 2) - 1; i++ {
    j = len(caractere) - 1 - i
    x = caractere[i]
    caractere[i] = caractere[j]
    caractere[j] = x
  }
  return string(caractere)
}

func Trocar(texto string, inicio, fim int, novoTexto string) string {
  var caractere = []rune(texto)
  caractere = append(caractere[:inicio], 
              append([]rune(novoTexto), 
              caractere[fim + 1:] ...) ...)
  return string(caractere)
}

func main() {
  var palavra string = "COMPUTADOR"

  fmt.Printf("Palavra ........: %s\n", palavra)
  fmt.Println()
  fmt.Printf("Lado esquerdo ..: %s\n", palavra[0:3])
  fmt.Printf("Centro .........: %s\n", palavra[3:7])
  fmt.Printf("Lado direito ...: %s\n", palavra[7:10])
  fmt.Printf("Reverso ........: %s\n", Inverte(palavra))
  fmt.Printf("Copias .........: %s\n", strings.Repeat(palavra, 3))
  fmt.Printf("Nova palavra ...: %s\n", Trocar(palavra, 6, 6, "O"))

  fmt.Println()
  fmt.Print("Tecle <Enter> para encerrar... ")
  fmt.Scanln()
}
